kong/api/routes/certificates.lua

```diff
@@ -3,20 +3,150 @@ local utils = require "kong.tools.utils"
 local cjson = require "cjson"
 
 
+local create_certificate = function(self, dao_factory, helpers)
+  local snis
+  if type(self.params.snis) == "string" then
+    snis = utils.split(self.params.snis, ",")
+    self.params.snis = nil
+  end
+
+  if snis then
+    -- dont add the certificate or any snis if we have an SNI conflict
+    -- its fairly inefficient that we have to loop twice over the datastore
+    -- but no support for OR queries means we gotsta!
+    local seen = {}
+
+    for _, sni in ipairs(snis) do
+      if seen[sni] then
+        return helpers.responses.send_HTTP_CONFLICT(
+          "duplicate requested sni name " .. sni
+        )
+      end
+
+      local cnt, err = dao_factory.ssl_servers_names:count({
+        name = sni,
+      })
+      if err then
+        return helpers.yield_error(err)
+      end
+
+      if cnt > 0 then
+        return helpers.responses.send_HTTP_CONFLICT(
+          "entry already exists with name " .. sni
+        )
+      end
+
+      seen[sni] = true
+    end
+  end
+
+  local ssl_cert, err = dao_factory.ssl_certificates:insert(self.params)
+  if err then
+    return helpers.yield_error(err)
+  end
+
+  -- insert SNIs if given
+
+  if snis then
+    ssl_cert.snis = {}
+
+    for _, sni in ipairs(snis) do
+      local ssl_server_name = {
+        name                = sni,
+        ssl_certificate_id  = ssl_cert.id,
+      }
+
+      local row, err = dao_factory.ssl_servers_names:insert(ssl_server_name)
+      if err then
+        return helpers.yield_error(err)
+      end
+
+      table.insert(ssl_cert.snis, row.name)
+    end
+  end
+
+  return helpers.responses.send_HTTP_CREATED(ssl_cert)
+end
+
 return {
   ["/certificates/"] = {
     POST = function(self, dao_factory, helpers)
+      create_certificate(self, dao_factory, helpers)
+    end,
+
+
+    GET = function(self, dao_factory, helpers)
+      local ssl_certificates, err = dao_factory.ssl_certificates:find_all()
+      if err then
+        return helpers.yield_error(err)
+      end
+
+      for i = 1, #ssl_certificates do
+        local rows, err = dao_factory.ssl_servers_names:find_all {
+          ssl_certificate_id = ssl_certificates[i].id
+        }
+        if err then
+          return helpers.yield_error(err)
+        end
+
+        ssl_certificates[i].snis = setmetatable({}, cjson.empty_array_mt)
+
+        for j = 1, #rows do
+          table.insert(ssl_certificates[i].snis, rows[j].name)
+        end
+      end
+
+      return helpers.responses.send_HTTP_OK({
+        data = #ssl_certificates > 0 and ssl_certificates or cjson.empty_array,
+        total = #ssl_certificates,
+      })
+    end,
+
+
+    PUT = function(self, dao_factory, helpers)
+
```

Comment:
style: could we avoid blank lines as such at the beginning of a function's body? Thanks!